### PR TITLE
Fast double to decimal cast

### PR DESCRIPTION
**Pull request by pnowojski**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java [hunk 1]
```diff
@@ -101,15 +101,15 @@ public void testDoubleToDecimalCasts()
         assertDecimalFunction("CAST(234.0 AS DECIMAL(4,1))", decimal("234.0"));
         assertDecimalFunction("CAST(.01 AS DECIMAL(3,3))", decimal(".010"));
         assertDecimalFunction("CAST(.0 AS DECIMAL(3,3))", decimal(".000"));
-        assertDecimalFunction("CAST(0 AS DECIMAL(1,0))", decimal("0"));
-        assertDecimalFunction("CAST(0 AS DECIMAL(4,0))", decimal("0000"));
-        assertDecimalFunction("CAST(1000 AS DECIMAL(4,0))", decimal("1000"));
+        assertDecimalFunction("CAST(0.0 AS DECIMAL(1,0))", decimal("0"));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java]: why did you change this?

Does `CAST(0 AS DECIMAL(8,0)` still work?

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java [hunk 2]
```diff
@@ -106,14 +106,33 @@ public void testDoubleToDecimalCasts()
         assertDecimalFunction("CAST(1000.0 AS DECIMAL(4,0))", decimal("1000"));
         assertDecimalFunction("CAST(1000.01 AS DECIMAL(7,2))", decimal("01000.01"));
         assertDecimalFunction("CAST(-234.0 AS DECIMAL(3,0))", decimal("-234"));
-        assertDecimalFunction("CAST(12345678901234567.0 AS DECIMAL(17,0))", decimal("12345678901234568"));
-        assertDecimalFunction("CAST(-12345678901234567.0 AS DECIMAL(17,0))", decimal("-12345678901234568"));
-        assertDecimalFunction("CAST(1234567890.0 AS DECIMAL(20,10))", decimal("1234567890.0000000000"));
-        assertDecimalFunction("CAST(-1234567890.0 AS DECIMAL(20,10))", decimal("-1234567890.0000000000"));
+        assertDecimalFunction("CAST(1234567890123456.0 AS DECIMAL(16,0))", decimal("1234567890123456"));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java]: can you add tests like
`assertDecimalFunction("CAST(1234567890123456.9 AS DECIMAL(16,0))", decimal("1234567890123457"));`
`assertDecimalFunction("CAST(-1234567890123456.9 AS DECIMAL(16,0))", decimal("-1234567890123457"));`
to test rounding?

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java [hunk 3]
```diff
@@ -106,14 +106,33 @@ public void testDoubleToDecimalCasts()
...
+        assertDecimalFunction("CAST(-1234567890123456.0 AS DECIMAL(16,0))", decimal("-1234567890123456"));
 
         assertInvalidCast("CAST(234.0 AS DECIMAL(2,0))", "Cannot cast DOUBLE '234.0' to DECIMAL(2, 0)");
         assertInvalidCast("CAST(1000.01 AS DECIMAL(5,2))", "Cannot cast DOUBLE '1000.01' to DECIMAL(5, 2)");
         assertInvalidCast("CAST(-234.0 AS DECIMAL(2,0))", "Cannot cast DOUBLE '-234.0' to DECIMAL(2, 0)");
+    }
+
+    @Test
+    public void testDoubleToLongDecimalCasts()
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java]: add some tests for rounding here as well

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 4]
```diff
@@ -464,6 +463,7 @@ public static long floatToShortDecimal(long value, long precision, long scale, l
     @UsedByGeneratedCode
     public static Slice floatToLongDecimal(long value, long precision, long scale, long tenToScale)
     {
+        // TODO: optimize
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java]: This comment is to general, nobody, but you (today), knows what can be optimized.
Can you please explain what and how can be optimized?

Or remove it.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 5]
```diff
@@ -90,6 +93,10 @@
             throw new RuntimeException(e);
         }
 
+        for (int i = 0; i < POWERS_OF_FIVE.length; ++i) {
+            POWERS_OF_FIVE[i] = unscaledDecimal(BigInteger.valueOf(5).pow(i));
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: Using multiply would be much faster here ;)

```
POWERS_OF_FIVE[0] = decimal(0)
POWERS_OF_FIVE[1] = decimal(5)
for (int i = 2; i < POWERS_OF_FIVE.length; ++i) {
     POWERS_OF_FIVE[i] = multiply(POWER_OF_FIVE[i-1], decimal(5));
}
```

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 6]
```diff
@@ -671,6 +801,134 @@ static void shiftRight(Slice decimal, int rightShifts, boolean roundUp, Slice re
...
+                        }
+                    }
+                }
+            }
+        }
+
+        pack(result, r0, r1, r2, r3, false);
+    }
+
+    // visible for testing
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: I think there is already an annotation for that

Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: http://google.github.io/guava/releases/19.0/api/docs/com/google/common/annotations/VisibleForTesting.html

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 7]
```diff
@@ -871,4 +1135,88 @@ private static void setRawLong(Slice decimal, int index, long value)
...
+                // 2**(twoScaleDown-scale)
+                // To check overflow while preserving precision, we need to do a
+                // real multiplication
+
+                multiplyShiftDestructive(unscaledValue, POWERS_OF_FIVE[scale], (twoScaleDown - scale));
+                //throw new NotImplementedException();
+                //this.unscaledValue.multiplyShiftDestructive(
+                //        SqlMathUtil.POWER_FIVES_INT128[scale],
+                //        (short) (twoScaleDown - scale));
+            }
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: remove this comment

File: presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java [hunk 8]
```diff
@@ -311,6 +396,53 @@ private static void assertShiftRight(Slice decimal, int rightShifts, boolean rou
...
     }
 
+    private void assertShiftRightArray8Overflow(BigInteger value, int rightShifts)
+    {
+        try {
+            assertShiftRightArray8(value, rightShifts);
+            fail();
+        }
+        catch (ArithmeticException ignored) {
+        }
```
Comment (kokosing) [inline presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java]: assert that exception is an overflow

File: presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java [hunk 9]
```diff
@@ -311,6 +396,53 @@ private static void assertShiftRight(Slice decimal, int rightShifts, boolean rou
...
+        assertEquals(decodeUnscaledValue(result), expectedResult);
+    }
+
+    private void assertShiftLeftOverflow(BigInteger value, int leftShifts)
+    {
+        try {
+            assertShiftLeft(value, leftShifts);
+            fail();
+        }
+        catch (ArithmeticException ignored) {
```
Comment (kokosing) [inline presto-spi/src/test/java/com/facebook/presto/spi/type/TestUnscaledDecimal128Arithmetic.java]: ditto

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 10]
```diff
@@ -1034,6 +1034,24 @@ public static void throwOverflowException()
         throw new ArithmeticException("Decimal overflow");
     }
 
+    public static int getInt(Slice decimal, int index)
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: please do method move in separate commits. Have this method changed (besides make it them public)?

**Comment by kokosing** [comment]
looks good, although I didn't review the formulas or bitwise operations

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java (hunk 1 above)
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java]: yes it works, but it wasn't `testDoubleToDecimalCasts`

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java (hunk 2 above)
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java]: done

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java (hunk 3 above)
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalCasts.java]: done

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java (hunk 4 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java]: done

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 5 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: but it doesn't matter, it's executed only once and `BigInteger` version is more reliable

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 6 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: but this is in `presto-spi` :(

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 7 above)
Comment (pnowojski) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: ops, fixed

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 11]
```diff
@@ -428,49 +430,61 @@ public static long longDecimalToFloat(Slice decimal, long precision, long scale,
...
     {
-        BigDecimal decimal = new BigDecimal(intBitsToFloat((int) value));
-        decimal = decimal.setScale((int) scale, ROUND_HALF_UP);
-        if (overflows(decimal, precision)) {
-            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast FLOAT '%s' to DECIMAL(%s, %s)", intBitsToFloat((int) value), precision, scale));
+        //TODO: optimize, implement specialized float to decimal conversion instead of using double to decimal
+        float floatValue = intBitsToFloat((int) value);
+        Slice decimal = UnscaledDecimal128Arithmetic.doubleToLongDecimal(floatValue, precision, (int) scale);
+        if (overflows(decimal, (int) precision)) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast FLOAT '%s' to DECIMAL(%s, %s)", floatValue, precision, scale));
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:485]: I know it's unrelated change but can you rename FLOAT to REAL since it's a leftover after renaming (there's no FLOAT anymore. New name for the same type is REAL)? It's fine to internally refer to it as float (as in method names) but any user-facing messages should refer to `REAL`.

File: presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java [hunk 12]
```diff
@@ -69,12 +70,13 @@
...
 import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
 import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.stream.Collectors.toList;
 import static java.util.stream.Collectors.toMap;
 import static org.openjdk.jmh.annotations.Scope.Thread;
 
 @State(Scope.Thread)
 @OutputTimeUnit(TimeUnit.MILLISECONDS)
-@Fork(value = 3, jvmArgs = "-XX:MaxInlineSize=100")
+@Fork(value = 1, jvmArgs = "-XX:MaxInlineSize=100")
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java:79]: Why?

File: presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java [hunk 13]
```diff
@@ -338,6 +385,31 @@ private RowExpression rowExpression(String expression)
...
+
+        private Object generateRandomValue(Type type)
+        {
+            if (type instanceof DoubleType) {
+                return random.nextDouble() * (2L * doubleMaxValue) - doubleMaxValue;
+            }
+            else if (type instanceof DecimalType) {
+                return randomDecimal((DecimalType) type);
+            }
+            else {
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java:397]: No need for `else`, just `throw` here.

**Comment by maciejgrzybek** [comment]
Is it safe to use polymorphic calls with `State` (https://github.com/pnowojski/presto/blob/46368232243d2608f7869b08512eea30fe5c99e7/presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java#L123) in benchmark? Isn't it too big room for Java optimizer which may affect test results?

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java (hunk 11 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java:485]: fixed

File: presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java (hunk 12 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/type/BenchmarkDecimalOperators.java:79]: fixed, unintended trash commit

**Comment by pnowojski** [comment]
@maciejgrzybek it's one virtual call per page, so it simply can not be significant. It's just as bad as the fact that Operators in Presto are polymorphic.
